✨ Generation of typed apply clients using upstream generator

JoelSpeed opened the pull request:
<!-- please add a icon to the title of this PR (see VERSIONING.md), and delete this line and similar ones -->
<!-- the icon will be either ⚠ (:warning:, major), ✨ (:sparkles:, minor), 🐛 (:bug:, patch), 📖 (:book:, docs), or 🏃 (:running:, other) -->

<!-- What does this do, and why do we need it? -->
This integrates the upstream applyconfig generator into controller-tools to allow generation of ApplyConfig style types from custom API types.

The aim here is to provide the pointer style structs that could then be passed to an Apply style method for using server side apply.

Setting this up as a WIP PR for now, current TODO list:
* [x] Work out if there's any extra stuff here that's not needed
* [x] Work out how to test this without the hack adding cronjob testdata to the go.mod
* [x] Do we need OpenAPI schema/do we need the extract functions?
  * Future work, after this PR has merged 
* [x] General review of code and cleanup - this was quite hacky when I started the branch
* [x] How would a user want this integrated into their project?
  * For now they can add the correct markers and it will generate as a subfolder of their API

JoelSpeed commented on pkg/applyconfigurations/gen.go:
```diff
@@ -0,0 +1,253 @@
+	"k8s.io/api/":                   "k8s.io/client-go/applyconfigurations/",
+}
+
+const importPathSuffix = "applyconfiguration"
```
Should this be configurable via a marker?

jakobmoellerdev replied:
This is now available as  a marker `kubebuilder:ac:output:package`

JoelSpeed commented on pkg/applyconfigurations/gen.go:
```diff
@@ -0,0 +1,253 @@
+	enableTypeMarker = markers.Must(markers.MakeDefinition("kubebuilder:ac:generate", markers.DescribesType, false))
+)
+
+var importMapping = map[string]string{
```
Should we allow additional import mappings, I know the upstream applyconfiguration gen does

JoelSpeed commented:
/test all

Need to work out how to test this, but want to see what the existing tests do with this so I can fix any failures introduced

JoelSpeed commented:
/test all

JoelSpeed commented on pkg/applyconfigurations/applyconfiguration_integration_test.go:
```diff
@@ -0,0 +1,144 @@
+		filesInMaster := make(map[string][]byte)
+		masterFileNames := sets.New[string]()
+		cronJobFS := os.DirFS(".")
+		masterPath := "applyconfiguration-master"
```
Because I can't change the directory to which we are generating the assets currently, this is what I came up with, create a copy that we keep checked in and then diff against that, if there are genuine changes we can update both copies.

Will look at gengo to see if there's a backwards compatible way to allow it to output to different places like `OutputRule` does for these generators

EraYaN commented:
As an option I would love to have the Extract functions, we use them extensively in our operator (for better or for worse).

sbueringer commented on pkg/applyconfigurations/gen.go:
```diff
@@ -0,0 +1,256 @@
+		return nil
+	}
+
+	genericArgs, _ := generatorargs.NewDefaults()
+	genericArgs.InputDirs = []string{root.PkgPath}
+	genericArgs.OutputPackagePath = filepath.Join(root.PkgPath, importPathSuffix)
+	genericArgs.GoHeaderFilePath = ctx.HeaderFilePath
```
If I see correctly this currently stores the generated files somewhere below GOPATH/src. Would be probably good to align this to how we can set output dir in other generators.

sbueringer replied:
I tried to run this in Cluster API with ~ `apply paths=./api/...` I end up with files generated in $GOPATH/src/sigs.k8s.io/cluster-api/api/v1beta1/applyconfiguration/api/v1beta1

sbueringer replied:
Not sure if it's good if we store the applyconfigurations in a sub-package of the API types. Might be a simple output:apply:dir would be better and then we can store it below there, e.g. 
* output:apply:dir is explicitly set to `GOPATH/src/sigs.k8s.io/cluster-api/applyconfigurations`
* files are then generated in e.g. `$GOPATH/src/sigs.k8s.io/cluster-api/applyconfiguration/api/v1beta1`

I think this would follow the same pattern as the ones in client-go. (side note: the API folders are not well structured in Cluster API)

sbueringer replied:
(probably this is covered by: https://github.com/kubernetes-sigs/controller-tools/pull/818#discussion_r1199118670 feel free to resolve)

sbueringer commented:
@JoelSpeed Can we get rid of the vendor folder? It's probably on my side but the performance of the file tab is pretty bad with 1700 files.

sbueringer commented on pkg/applyconfigurations/testdata/cronjob/applyconfiguration-master/internal/internal.go:
```diff
@@ -0,0 +1,46 @@
+// Code generated by applyconfiguration-gen. DO NOT EDIT.
+
+package internal
```
Q: Do you know why this is generated?

sbueringer replied:
Thx!

alvaroaleman commented:
> @JoelSpeed Can we get rid of the vendor folder? It's probably on my side but the performance of the file tab is pretty bad with 1700 files.

Second this, I can literally not review this through the GitHub UI

alvaroaleman reviewed (commented):
For an internal project, I added the `+kubebuilder:ac:generate=true` comment both on the package and on my type, running `controller-gen` with `apply paths=./pkg/api/v1/` succeeds but doesn't generate anything - any idea what I am am doing wrong?

JoelSpeed commented:
@jakobmoellerdev Reached out to me last week and sent me a patch that we think resolves a number of the comments here. 

@jakobmoellerdev Anything the reviewers need to know about what you've changed?

jakobmoellerdev commented on pkg/applyconfigurations/gen.go:
```diff
@@ -0,0 +1,238 @@
+	// looking into the package and retrieving the first go file it finds, and using that as the output base.
+	// this is because we cannot rely on gogen calculating the correct output base.
+	// if we leave this empty, gogen will attempt to use GOPATH to write the files which is not wanted
+	genericArgs.OutputBase = filepath.Dir(root.GoFiles[0])
```
I think with some more invest it may be possible to resolve the directory smoother but there we issues with the output package not having correct import paths. For now this will work fine with the generation of the base directory being in the package path. For the request of having a directory outside of the package path this needs adjustment though. Open to any ideas here

jakobmoellerdev replied:
I believe in general it should now work both within and outside GOPATH and should resolve the issue @sbueringer and @alvaroaleman not seeing the generated files in the right folders

jakobmoellerdev commented on pkg/oot_testsubdirectory:
I believe this file was included due to the patch, should be removed from the final version

erikgb commented:
@JoelSpeed Do you have some bandwidth to follow up on this PR? I just added the upstream generator to one of our projects to get some progress towards SSA, and it was NOT a pleasant experience! I would love to see the ability to generate apply clients with controller-gen.

JoelSpeed commented:
@erikgb I'm currently waiting on a second round of review from those who previously reviewed. @jakobmoellerdev kindly continued the work and we pushed his updates to my branch, but haven't had a review since then

jakobmoellerdev commented:
New year new bump :) any chance this can get another round of review?

jakobmoellerdev commented:
@sbueringer Bringing this back from the depths of your PR pile!

sbueringer commented:
Not sure if I actually asked it somewhere :). @JoelSpeed How fundamentally might this PR change with the rebase that is necessary for 1.30?

JoelSpeed commented:
> How fundamentally might this PR change with the rebase that is necessary for 1.30?

Great question. I know there's been a massive rewrite of the gengo stuff, I have some WIP stuff in place for some OpenShift tooling that was affected by that, but in general, it wasn't massive. I would hope this would be similar, but yes, we need to get the deps updated to 1.30 and see tbh

sbueringer commented:
Was just wondering if it makes sense for me to review the PR before the rebase. I could commit to actually doing the review assuming you find the time for a rebase :)

jakobmoellerdev commented:
@JoelSpeed I take that as you choose to do the rebase or I submit another patch. :) PLMK

JoelSpeed commented:
I likely won't have time until next friday to take that on, but if you want to have a go @jakobmoellerdev I can force push whatever you come up with over my branch again if you wanted?

EraYaN commented:
This is still relevant

/remove-lifecycle stale

erikgb commented:
Still highly relevant!

/remove-lifecycle stale

JoelSpeed commented:
Right, lets try and work out what the minimum amount is that we need to do here to get this over the line

From my initial comment, I had two things in mind that needed to be answered

> Do we need OpenAPI schema/do we need the extract functions?

If we don't have the extract functions, are we confident that this is still useful? IMO, yes. In most of the use cases I've seen for applyconfig/SSA so far, they haven't actually used the extract functions (@mdbooth keep me honest here, you have projects I'm thinking of)

So for now, I think we can say that the extract functions that would normally be generated from the openapi schema are optional, and we can work on those separately

> How would a user want this integrated into their project?

Do we want to generate this automatically? Do we want to choose a location for the generated files? I guess that isn't necessarily a question for this PR since it will be sorted in the kubebuilder repo instead right?

---

Other than rebasing then, what's actually left to sort here? I need to go through @jakobmoellerdev's outstanding comments I guess

sbueringer commented:
> Do we want to generate this automatically? Do we want to choose a location for the generated files? I guess that isn't necessarily a question for this PR since it will be sorted in the kubebuilder repo instead right?

Would be good to have an idea how it's usually used (I have no experience with apply configurations). Are there some common patterns?

Overall sounds fine to me. I would take a closer look once it's rebased

JoelSpeed commented:
> Would be good to have an idea how it's usually used (I have no experience with apply configurations). Are there some common patterns?

So there are two ways you can use apply configurations generally. 

1. Extract and modify
  * In this case, you use the extract functions that we aren't generating to extract an ApplyConfiguration from the regular type
  * The extract function looks at the managed fields to determine which fields you set previously, and provides the object you last applied.
  * You can then make modifications to this, and apply again
2. From scratch
  * On each reconcile you create the minimal apply configuration (metadata)
  * As you pass through your reconcile function, you need to set every field you care about
  * Eg, if you normally set 4 conditions, you must set all 4 conditions through every reconcile loop
  * If you do not set all fields (say you exit early and forget to set one condition), then whatever you didn't set is removed on the next update
  * I have seen examples of people combining this approach with the traditional object to ensure things like lastTransitionTime is correct on condiitons

In a recent example in my day to day, we are using SSA and apply configs to apply conditions from several control loops to a single object. Each loop has two conditions it cares about, so we just make sure that those conditions are set correctly on each loop. We merge the `reconcile` output of conditions with the previous conditions that we set (merge meaning setting the transition times correctly) just before we send the apply.

I'll admit, I'm not the most heavy user of these myself so would be good to get other's opinions/experiences too

sbueringer commented:
Thx for the additional context! I was mostly wondering where folks are generating the files to :)

JoelSpeed commented:
For us, we generate them and place them alongside our generated clients/listers/informers for our types. We do keep those in a separate repo to our APIs though and mimic what K8s does with API vs client-go repos, so perhaps not the best case study

alvaroaleman commented:
> So for now, I think we can say that the extract functions that would normally be generated from the openapi schema are optional, and we can work on those separately

Yeah IMHO having the apply configs would be a big step forward. I am unsure about how needed or not the extract functions are, but worst case we can add that later on.

> Do we want to generate this automatically? Do we want to choose a location for the generated files? I guess that isn't necessarily a question for this PR since it will be sorted in the kubebuilder repo instead right?

Not a strong opinion but maybe just generate it if an output path is configured or such? I am not completely sure if we want this by default. We can still do it by default later on if we find that makes sense but once we did that, we can not go back

EraYaN commented:
The Extract functions are very useful to create easy to use controllers. I'd very much like them to be included, and if you start running the tools you might as well run those to get the openapi spec.

We internally often generate the applyconfigurations in a different folder, so if the models are in `api/v1` the applyconfigurations are in `client/applyconfigurations/v1`. Normal consumers probably don't need them, (although we do consume them in our internal kubectl plugin/CLI)

sbueringer commented:
For me it's not required that a first PR has to implement everything including Extract functions. I think a lot of folks would be happy with seeing any sort of progress here.

EraYaN commented:
I mean you can probably nick the little tool you can use to get the OpenAPI spec (1 file of go code) from just about any other project that does this, you really just have to marshall the generated object to JSON. It's one more call after running openapi-gen to get the actual json out. openapi-gen should really do this by itself but alas.

<details>
<summary>Example (that we borrowed from cert-manager)</summary>

```go
/*
Copyright 2022 The cert-manager Authors.
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at
    http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

/*
Copyright 2022 The Kubernetes Authors.
Started from https://github.com/kubernetes/kubernetes/blob/f5956716e3a92fba30c81635c68187653f7567c2/pkg/generated/openapi/cmd/models-schema/main.go
*/

package main

import (
	"encoding/json"
	"fmt"
	"os"
	"strings"

	"k8s.io/apiextensions-apiserver/pkg/generated/openapi"
	"k8s.io/kube-openapi/pkg/common"
	"k8s.io/kube-openapi/pkg/validation/spec"
)

// Outputs openAPI schema JSON containing the schema definitions in zz_generated.openapi.go.
func main() {
	err := output()
	if err != nil {
		os.Stderr.WriteString(fmt.Sprintf("Failed: %v", err))
		os.Exit(1)
	}
}

func output() error {
	refFunc := func(name string) spec.Ref {
		return spec.MustCreateRef(fmt.Sprintf("#/definitions/%s", friendlyName(name)))
	}
	defs := openapi.GetOpenAPIDefinitions(refFunc)
	//openapi.Hack__AddGatewayApiOpenAPIDefinitions(defs, refFunc)

	schemaDefs := make(map[string]spec.Schema, len(defs))
	for k, v := range defs {
		// Replace top-level schema with v2 if a v2 schema is embedded
		// so that the output of this program is always in OpenAPI v2.
		// This is done by looking up an extension that marks the embedded v2
		// schema, and, if the v2 schema is found, make it the resulting schema for
		// the type.
		if schema, ok := v.Schema.Extensions[common.ExtensionV2Schema]; ok {
			if v2Schema, isOpenAPISchema := schema.(spec.Schema); isOpenAPISchema {
				schemaDefs[friendlyName(k)] = v2Schema
				continue
			}
		}

		schemaDefs[friendlyName(k)] = v.Schema
	}
	data, err := json.Marshal(&spec.Swagger{
		SwaggerProps: spec.SwaggerProps{
			Definitions: schemaDefs,
			Info: &spec.Info{
				InfoProps: spec.InfoProps{
					Title:   "calcasa-operator",
					Version: "unversioned",
				},
			},
			Swagger: "2.0",
		},
	})
	if err != nil {
		return fmt.Errorf("error serializing api definitions: %w", err)
	}
	os.Stdout.Write(data)
	return nil
}

// From vendor/k8s.io/apiserver/pkg/endpoints/openapi/openapi.go
func friendlyName(name string) string {
	nameParts := strings.Split(name, "/")
	// Reverse first part. e.g., io.k8s... instead of k8s.io...
	if len(nameParts) > 0 && strings.Contains(nameParts[0], ".") {
		parts := strings.Split(nameParts[0], ".")
		for i, j := 0, len(parts)-1; i < j; i, j = i+1, j-1 {
			parts[i], parts[j] = parts[j], parts[i]
		}
		nameParts[0] = strings.Join(parts, ".")
	}
	return strings.Join(nameParts, ".")
}
```

and we run it like this
```makefile
.PHONY: generate-openapi
generate-openapi: $(DEPENDS_ON_GO)
	@go run hack/crd-schema/main.go
```

and that gets piped to a file.
</details>

JoelSpeed commented:
@EraYaN The issue is, we want to provide a "seamless" experience for the users of these projects. They shouldn't need to know anything about generating the openapi schema or why it's required.

You're right, it is one little go file that's required, but in terms of what's required for a kubebuilder project, that means implementing a generator that generates that file, and then we have to include the scripts etc to get that running, at the right time, and including options of where to put it, and then exposing options on how to pass it to the apply config generator too if it's in a custom location.

The work to do this for one repo is fairly quick, to be able to generate the correct thing for kubebuilder projects, less simple.

EraYaN commented:
`openapi-gen` already generates a normal source file, it could use the same location I suppose. But I recognize it might just be a step too far. It's a real shame the upstream generators are so hard to use.

alvaroaleman commented:
> openapi-gen already generates a normal source file, it could use the same location I suppose. But I recognize it might just be a step too far. It's a real shame the upstream generators are so hard to use.

Yeah - Lets focus on getting something done over perfection here, please

JoelSpeed commented on pkg/applyconfigurations/applyconfiguration_integration_test.go:
```diff
@@ -0,0 +1,151 @@
+		By("initializing the runtime")
+		optionsRegistry := &markers.Registry{}
+		Expect(optionsRegistry.Register(markers.Must(markers.MakeDefinition("crd", markers.DescribesPackage, crd.Generator{})))).To(Succeed())
+		// Add the applyconfigurations generator but set it to verify only.
+		// This allows us to check if there's a diff between the checked in generated data and what it would generate now.
```
Is this correct? We seem to make the output var and set it on output rules, but I don't think that's actually being used later 🤔

sbueringer replied:
I'm not following :)

JoelSpeed commented on pkg/applyconfigurations/gen.go:
```diff
@@ -0,0 +1,230 @@
+		outpkg = importPathSuffix
+	}
+
+	arguments.OutputDir = filepath.Join(root.Dir, outpkg)
+	arguments.OutputPkg = filepath.Join(root.Package.PkgPath, outpkg)
```
At the moment, we assume that the outpkg is a subdirectory relative to the current package directory. In the future we may want to make it so that you can provide a full package

If anyone has a good idea on how to detect `sigs.k8s.io/....` vs just a `applyconfiguration` for being a real package or not, open to fixing before we merge

JoelSpeed commented:
@alvaroaleman @sbueringer Would you mind adding this to your review queue 🙏

sbueringer commented on pkg/applyconfigurations/gen.go:
```diff
@@ -0,0 +1,230 @@
+		enableTypeMarker, markers.SimpleHelp("apply", "overrides enabling or disabling applyconfigurations generation for the type"))
+	into.AddHelp(
+		enablePkgMarker, markers.SimpleHelp("apply", "overrides enabling or disabling applyconfigurations generation for the package"))
+	into.AddHelp(
```
should we mention the default value?

sbueringer commented on pkg/applyconfigurations/gen.go:
```diff
@@ -0,0 +1,230 @@
+
+	into.AddHelp(isCRDMarker,
+		markers.SimpleHelp("apply", "enables apply configuration generation for this type"))
+	into.AddHelp(
+		enableTypeMarker, markers.SimpleHelp("apply", "overrides enabling or disabling applyconfigurations generation for the type"))
+	into.AddHelp(
+		enablePkgMarker, markers.SimpleHelp("apply", "overrides enabling or disabling applyconfigurations generation for the package"))
```
Let's explain a bit how these markers can be used in combination to enable/disable generation for a pkg/type

sbueringer commented on pkg/applyconfigurations/gen.go:
```diff
@@ -0,0 +1,230 @@
+	return false, nil
+}
+
+// enableOnType marks whether applyconfiguration generation is enabled for the type.
```
nit: "marks" might be the wrong verb

(I'm also fine with no godoc comment here, the func name is good enough)

Same in l.115

sbueringer commented on pkg/applyconfigurations/gen.go:
```diff
@@ -0,0 +1,230 @@
+		if err != nil {
+			return fmt.Errorf("failed to create temporary file: %w", err)
+		}
+		tmpFile.Close()
```
let's handle the error

sbueringer commented on pkg/applyconfigurations/gen.go:
```diff
@@ -0,0 +1,230 @@
+	}
+
+	for _, pkg := range ctx.Roots {
+		if err := objGenCtx.generateForPackage(pkg); err != nil {
```
Would it make sense to try to generate as much as possible and then only returning all errors at the end?

sbueringer commented on pkg/applyconfigurations/gen.go:
```diff
@@ -0,0 +1,230 @@
+
+// generateForPackage generates apply configuration implementations for
+// types in the given package, writing the formatted result to given writer.
+// May return nil if source could not be generated.
```
What does that mean? :) That we might not get an error if "source could not be generated"? (what is source in this context?)

sbueringer commented on pkg/applyconfigurations/gen.go:
```diff
@@ -0,0 +1,230 @@
+
+	outpkg, _ := outputPkg(ctx.Collector, root)
+	if outpkg == "" {
+		outpkg = importPathSuffix
```
nit: Struggling a bit with importPathSuffix, should this be "defaultOutputPackage"?

(maybe consider moving the defaulting into the outputPkg func)

sbueringer commented on pkg/applyconfigurations/testdata/cronjob/applyconfiguration/internal/internal.go:
```diff
@@ -0,0 +1,46 @@
+// Code generated by applyconfigurations. DO NOT EDIT.
```
Having applyconfiguration and testapplyconfiguration looks like have the golden files twice?

Do we have to check these in twice? Or can we e.g. get away with having one of these folders on .gitignore?

sbueringer commented on pkg/applyconfigurations/testdata/cronjob/cronjob_types.go:
```diff
@@ -0,0 +1,174 @@
+/*
```
Would it make sense to sync this with pkg/crd/testdata/cronjob_types.go?

I think the one in the CRD folder has a few more fields nowadays

sbueringer commented on pkg/applyconfigurations/testdata/cronjob/go.mod:
```diff
@@ -0,0 +1,50 @@
+
+require (
+	k8s.io/api v0.32.0
+	k8s.io/apimachinery v0.32.0
```
Maybe worth syncing with the top-level go.mod

(we have so many go.mod in this repo and I think none of them are in sync / regularly bumped. I think we have to get them all to the latest version, but not in this PR of course)

sbueringer commented on pkg/applyconfigurations/testdata/cronjob/groupversion_info.go:
```diff
@@ -0,0 +1,55 @@
+	GroupName     = "testdata.kubebuilder.io"
+	GroupVersion  = schema.GroupVersion{Group: GroupName, Version: "v1"}
+	schemeBuilder = runtime.NewSchemeBuilder(addKnownTypes)
+	// Install is a function which adds this version to a scheme
+	Install = schemeBuilder.AddToScheme
+
+	// SchemeGroupVersion generated code relies on this name
+	// Deprecated
+	SchemeGroupVersion = GroupVersion
+	// AddToScheme exists solely to keep the old generators creating valid code
+	// DEPRECATED
+	AddToScheme = schemeBuilder.AddToScheme
```
Is this the current state of the art? Install and AddToScheme deprecated?

JoelSpeed replied:
I believe this came from a Kubebuilder generated project (which may have been generated some time ago), will work out what we actually need

sbueringer commented on pkg/applyconfigurations/testdata/cronjob/groupversion_info.go:
```diff
@@ -0,0 +1,55 @@
+	AddToScheme = schemeBuilder.AddToScheme
+)
+
+// Resource generated code relies on this being here, but it logically belongs to the group
+// DEPRECATED
```
Does this mean the applyconfigurations need this? Can we clarify what DEPRECATED means for us in this context?

(this func is unused, so I guess we don't really need it?)

sbueringer commented on pkg/applyconfigurations/testdata/cronjob/applyconfiguration/utils.go:
```diff
@@ -0,0 +1,37 @@
+		return &testdatacronjob.CronJobSpecApplyConfiguration{}
+	case cronjob.SchemeGroupVersion.WithKind("CronJobStatus"):
+		return &testdatacronjob.CronJobStatusApplyConfiguration{}
+	case cronjob.SchemeGroupVersion.WithKind("ExampleStruct"):
+		return &testdatacronjob.ExampleStructApplyConfiguration{}
```
Is this intended? (same for AssociativeType and *Spec, *Status)

sbueringer commented on pkg/applyconfigurations/testdata/cronjob/applyconfiguration/utils.go:
```diff
@@ -0,0 +1,37 @@
+
+// ForKind returns an apply configuration type for the given GroupVersionKind, or nil if no
+// apply configuration type exists for the given GroupVersionKind.
+func ForKind(kind schema.GroupVersionKind) interface{} {
```
Is there any way we could have some unit test that tries to use this code and does some minimal verification it "works"?

sbueringer reviewed (commented):
Thank you very much!

Nothing major from my side

sbueringer commented:
@JoelSpeed Might be interesting to open a PR against Cluster API, configure the generator and check if everything behaves as expected

sbueringer commented on pkg/applyconfigurations/testdata/cronjob/go.mod:
```diff
@@ -0,0 +1,50 @@
+module sigs.k8s.io/controller-tools/pkg/applyconfigurations/testdata/cronjob
```
Looking at all these dependencies. If I would have a separate module for the API to only have minimal dependencies, I probably wouldn't put the applicationconfigurations in the API folder.

Does the current configuration allow to put the applyconfigurations in an arbitrary folder? (not below the package for which they are generated)

JoelSpeed replied:
This would be a subpackage of the API, so Go is smart enough not to import the applyconfigurations subpackage if you aren't using it, however, it is a good point.

I was playing with this last week and I believe you can put a fully qualified package name in, will double check, and perhaps demonstrate that through a test somehow

JoelSpeed commented on cmd/controller-gen/main.go:
```diff
@@ -51,6 +53,7 @@ var (
 		"crd":         crd.Generator{},
 		"rbac":        rbac.Generator{},
 		"object":      deepcopy.Generator{},
+		"apply":       applyconfigurations.Generator{},
```
Yeah that makes sense to me, the conflict with `kubectl apply` could be confusing

JoelSpeed commented on pkg/applyconfigurations/applyconfiguration_integration_test.go:
```diff
@@ -0,0 +1,157 @@
+	masterPath := "applyconfiguration"
+	By("resetting the test directory")
+	// not using after each to allow for inspection of failed tests
+	BeforeEach(func() {
```
Looking at the way this file is written, I think I picked this up from someone else's commit (doesn't look like my style), so will need to investigate this one a bit further to understand what the intention was

JoelSpeed commented on pkg/applyconfigurations/applyconfiguration_integration_test.go:
```diff
@@ -0,0 +1,157 @@
+		})).To(Succeed())
+
+		// Every file should be in both sets, check for files not in both sets.
+		Expect(masterFileNames.SymmetricDifference(outputFileNames).UnsortedList()).To(BeEmpty(), "Generated files should match the checked in files")
```
Yeah agreed, will update

sbueringer commented:
Just a heads up. CT main is now using v0.33 alpha3 + Go 1.24. Probably makes sense to rebase

JoelSpeed commented:
Had to rebase to resolve a conflict, the interesting/reviewable changes since last time start with https://github.com/kubernetes-sigs/controller-tools/pull/818/commits/79252fb20603335e5ffee6f32f4f1eaebc0fc0c8

I've addressed the feedback from Stefan, and included a major rewrite of the tests to avoid checking in the files twice, and to show that we can generate the applyconfig in a relative path outside of the API directory

sbueringer commented on pkg/applyconfiguration/applyconfiguration_integration_test.go:
```diff
@@ -0,0 +1,217 @@
+	return nil
+}
+
+var _ = Describe("CRD Generation From Parsing to CustomResourceDefinition", func() {
```
We should probably change this text

sbueringer commented on pkg/applyconfiguration/applyconfiguration_integration_test.go:
```diff
@@ -0,0 +1,217 @@
+			// Make sure the import paths are correct for the newly generated content.
+			content = []byte(strings.ReplaceAll(string(content), "testdata/cronjob/api/v1/applyconfiguration", filepath.Join("testdata/cronjob/api/v1", outputPackage)))
+
+			Expect(string(filesInOutput[name])).To(Equal(string(content)), "Generated files should match the checked in files, diff found in %s", name)
```
```suggestion
			Expect(string(filesInOutput[name])).To(BeComparableTo(string(content)), "Generated files should match the checked in files, diff found in %s", name)
```
Up to you, usually this diff is better

sbueringer commented on pkg/applyconfiguration/testdata/cronjob/go.mod:
```diff
@@ -0,0 +1,51 @@
+	sigs.k8s.io/yaml v1.4.0 // indirect
+)
+
+replace sigs.k8s.io/controller-tools => ../../../../
```
I think this replace is not needed

sbueringer commented on pkg/applyconfiguration/testdata/cronjob/api/v1/groupversion_info.go:
```diff
@@ -0,0 +1,43 @@
+// +versionName=v1
+// +kubebuilder:ac:generate=true
+// +kubebuilder:ac:output:package="applyconfiguration"
+package testdata
```
Probably doesn't really matter, but maybe let's change this to `package v1`

sbueringer commented:
/assign @alvaroaleman @vincepri 

Pending the few small findings above this is ready to merge from my side

sbueringer commented:
Thx!

/lgtm

alvaroaleman reviewed (approved):
Looks good, thank you so much for your work on this!
